repub_test.go

```diff
@@ -66,10 +69,78 @@ func TestRepublisher(t *testing.T) {
 		t.Fatal("waited too long for pub!")
 	}
 
-	cancel()
+	close(stopUpdates)
 
-	err := rp.Close()
+	// Check that republishing update does not call pubfunc again
+	rp.Update(testCid2)
+	err := rp.WaitPub(ctx)
 	if err != nil {
-		t.Fatal(err)
+		t.Error("got error trying to republish update:", err)
+	}
+	select {
+	case <-pub:
+		t.Fatal("pub func called again with repeated update")
+	case <-time.After(tlong / 2):
+	}
+
+	ctx3, cancel := context.WithTimeout(context.Background(), time.Millisecond)
+	defer cancel()
+
+	// Check that waitpub times out when blocked pubfunc is called
+	rp.Update(testCid1)
+	err = rp.WaitPub(ctx3)
+	if err != context.DeadlineExceeded {
+		t.Errorf("expected %q, got %v", context.DeadlineExceeded, err)
+	}
+
+	stopPub := make(chan struct{})
+	stopped := make(chan struct{})
+	go func() {
+		defer close(stopped)
+		for {
+			select {
+			case <-pub:
+			case <-stopPub:
+				return
+			}
+		}
+	}()
```

Comment:
I'm not following the objective here, why do we need this? At this point it seems we don't interact with `pub` anymore (and we stopped the continuous updates with `stopUpdates`), if we need a publish function that doesn't block we should redefine it here to make this more explicit and reduce complexity (but maybe I'm missing the point and that was not the use of this goroutine).

Reply:
This is necessary to continue clearing the `pub` channel.  Otherwise, the republisher's call to `pubfunc` will never return, and calling `Close()` will timeout.  The `stopped` channel is however not needed.